pkg/driver/trace_driver.go

```diff
@@ -100,13 +103,217 @@ func (d *Driver) runCSVWriter(records chan interface{}, filename string, writerD
+		}
+		for i := 0; i < len(DAGNode.Value.(*common.Node).Branches); i++ {
+			nodeQueue = append(nodeQueue, dummyNode)
+			copy(nodeQueue[1:], nodeQueue)
+			nodeQueue[0] = DAGNode.Value.(*common.Node).Branches[i].Front()
+		}
+		if DAGNode.Next() == nil {
+			println(printMessage)
+			buffer = ""
+			if len(nodeQueue) > 0 {
+				startingNode = true
+			} else {
+				break
+			}
+		} else {
+			nodeQueue = append(nodeQueue, dummyNode)
+			copy(nodeQueue[1:], nodeQueue)
+			nodeQueue[0] = DAGNode.Next()
+		}
+	}
+
+}
+
+func getMaxInvocation(functionList []*common.Function) []int {
+	maxInvocation := make([]int, len(functionList[0].InvocationStats.Invocations))
+	for _, i := range functionList {
+		for index, invocation := range i.InvocationStats.Invocations {
+			maxInvocation[index] = max(maxInvocation[index], invocation)
+		}
+	}
+	return maxInvocation
+
+}
+
+func getName(function *common.Function) int {
+	parts := strings.Split(function.Name, "-")
+	if parts[0] == "test" {
+		return 0
+	}
+	functionId, err := strconv.Atoi(parts[2])
+	if err != nil {
+		log.Fatal(err)
+	}
+	return functionId
+}
+
+// Read the Cumulative Distribution Frequency (CDF) of the widths and depths of a DAG
+func generateCDF(file string) [][]float64 {
+	sheetName := "data"
+	f, err := excelize.OpenFile(file)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			fmt.Println(err)
+		}
+	}()
+	// Removing the first 2 header rows
+	_ = f.RemoveRow(sheetName, 1)
+	_ = f.RemoveRow(sheetName, 1)
+	cols, _ := f.GetCols(sheetName)
+
+	cdf := make([][]float64, len(cols))
+	for i := range cols {
+		cdf[i] = make([]float64, len(cols[i]))
+	}
+	for i := 0; i < 6; i += 2 {
+		for j := 0; j < len(cols[i]); j++ {
+			cdfProb, _ := strconv.ParseFloat(strings.TrimSuffix(cols[i+1][j], "%"), 64)
+			cdfValue, _ := strconv.ParseFloat(cols[i][j], 64)
+			cdfProb = math.Round(cdfProb*100) / 100
+			cdf[i+1][j] = cdfProb
+			cdf[i][j] = cdfValue
+			if cdfProb == 100.00 {
+				cdf[i] = cdf[i][:j+1]
+				cdf[i+1] = cdf[i+1][:j+1]
+				break
+			}
+
+		}
+	}
+	return cdf
+}
+
+// Generate pseudo-random probabilities and compare it with the given CDF to obtain the depth and width of the DAG
+func getDAGStats(cdf [][]float64, maxSize int) (int, int) {
+	var width int
+	var depth int
+	depthProb := math.Round((rand.Float64() * 10000)) / 100
+	widthProb := math.Round((rand.Float64() * 10000)) / 100
+	for i, value := range cdf[1] {
+		if value == widthProb {
+			width = int(cdf[0][i])
+			break
+		}
+		if value > widthProb {
+			width = int(cdf[0][i-1])
+			break
+		}
+	}
+	for i, value := range cdf[3] {
+		if value == depthProb {
+			depth = int(cdf[2][i])
+			break
+		}
+		if value > depthProb {
+			depth = int(cdf[2][i-1])
+			break
+		}
+	}
+	// Re-run DAG Generation if size exceeds number of functions
+	if maxSize < width*(depth-1) {
+		width, depth = getDAGStats(cdf, maxSize)
+	}
+	return width, depth
```

Comment:
If maxSize is very small, this code segment has the potential for an infinite loop. Could it be modified to iterate a certain number of times instead? For example, after 10 attempts, if it still can't generate valid width and depth values, it could return a 1 * maxSize DAG.

Reply:
Noted! Will work on this